api/v1alpha1/virtualmachinepublishrequest_types.go

```diff
@@ -0,0 +1,530 @@
+	// the VM is published, as well as whether or not to overwrite an
+	// existing object with the same name.
+	//
+	// Please note this value is optional and if omitted, the controller
+	// will use spec.source.name + "-image" as the name of the published
+	// item.
+	//
+	// +optional
+	Item VirtualMachinePublishRequestTargetItem `json:"item,omitempty"`
+
+	// Location contains information about the location to which to publish
+	// the VM.
+	//
+	// +optional
+	Location VirtualMachinePublishRequestTargetLocation `json:"location,omitempty"`
+}
+
+// VirtualMachinePublishRequestSpec defines the desired state of a
+// VirtualMachinePublishRequest.
+//
+// All the fields in this spec are optional. This is especially useful when a
+// DevOps persona wants to publish a VM without doing anything more than
+// applying a VirtualMachinePublishRequest resource that has the same name
+// as said VM in the same namespace as said VM.
+type VirtualMachinePublishRequestSpec struct {
+	// Source is the source of the publication request, ex. a VirtualMachine
+	// resource.
+	//
+	// If this value is omitted then the publication controller checks to
+	// see if there is a resource with the same name as this
+	// VirtualMachinePublishRequest resource, an API version equal to
+	// spec.source.apiVersion, and a kind equal to spec.source.kind. If such
+	// a resource exists, then it is the source of the publication.
+	//
+	// +optional
+	Source VirtualMachinePublishRequestSource `json:"source,omitempty"`
+
+	// Target is the target of the publication request, ex. item
+	// information and a ContentLibrary resource.
+	//
+	// If this value is omitted, the controller uses spec.source.name + "-image"
+	// as the name of the published item. Addtionally, when omitted the
+	// controller attempts to identify the target location by matching a
+	// resource with an API version equal to spec.target.location.apiVersion, a
+	// kind equal to spec.target.location.kind, w/ the label
+	// "imageregistry.vmware.com/default".
+	//
+	// Please note that while optional, if a VirtualMachinePublishRequest sans
+	// target information is applied to a namespace without a default
+	// publication target, then the VirtualMachinePublishRequest resource
+	// will be marked in error.
+	//
+	// +optional
+	Target VirtualMachinePublishRequestTarget `json:"target,omitempty"`
+
+	// TTLSecondsAfterFinished is the time-to-live duration for how long this
+	// resource will be allowed to exist once the publication operation
+	// completes. After the TTL expires, the resource will be automatically
+	// deleted without the user having to take any direct action.
+	//
+	// If this field is unset then the request resource will not be
+	// automatically deleted. If this field is set to zero then the request
+	// resource is eligible for deletion immediately after it finishes.
+	//
+	// +optional
+	TTLSecondsAfterFinished *int64 `json:"ttlSecondsAfterFinished,omitempty"`
```

Comment:
I wonder if K8s would set the value to 0 if it's not provided and therefore cause the unexpected deletion immediately?

Reply:
Oh nvm! I saw it's a pointer type now.

Reply:
One more thing to clarify here, by saying "exist once the publication operation _completes_", does it mean the auto-deletion only occurs when the publish request completes successfully (i.e. after reaching its "Complete" condition)?

Reply:
Completes doesn't connote success or fail. Ready is only true if all conditions are met. If the publication ever fails terminally then complete is true, regardless of success.